### DISCUSSION TITLE
How does toga uses open_ File_ Dialog() on Android？

### DISCUSSION THREAD

**Discussion by wanglifeng95** [beeware/toga#1990] [Q&A] [answered]
The app uses   main_window.open_file_dialog()   to open files, which works normally on Ubuntu and Windows, but does not work on Android. The following prompt appears：
python.stdout: [Android] Not implemented: Window.open_file_dialog()

Why is this and how to solve it？

**Comment by mhsmith** [comment]
The Toga file dialog API isn't implemented on Android yet, but you can use the native APIs directly as shown at https://github.com/beeware/toga/pull/1158#issuecomment-1446196973.

**Reply by wanglifeng95** [reply]
Thanks for your help. I am glad that I did open the file dialog box on Android using the code. But the following error occurred:
python.stderr: AttributeError: 'ParcelFileDescriptor$AutoCloseInputStream' object has no attribute 'readAllBytes'

The code is as follows：
```
async def openFileDialog(self, widget):
    from toga.platform import current_platform
    print("current_platform : ", current_platform)
    if current_platform == "android":
        from android.content import Intent

        fileChose = Intent(Intent.ACTION_GET_CONTENT)
        fileChose.addCategory(Intent.CATEGORY_OPENABLE)
        fileChose.setType("*/*")

        # Assuming `app` is your toga.App object
        results = await app._impl.intent_result(Intent.createChooser(fileChose, "Choose a file"))
        data = results['resultData'].getData()
        context = app._impl.native
        bytesJarray = bytes((context.getContentResolver().openInputStream(data).readAllBytes()))
        # will do something to get filePath and return filePath
    else:
        filePath = await config.app.main_window.open_file_dialog("打开文件")
    return filePath
```

1.  On Android, the above code can open a file dialog, but how do I get the path of the selected file?
2.  When I try to read a file in a folder such as the Download folder on Android, a permission error occurs：
python.stdout: PermissionError: [Errno 13] Permission denied: '/storage/emulated/0/Download/you.jpg‘

In Kivy, it can be done by adding the following code, but it does not work in Toga
```
from android.permissions import request_permissions, Permission
request_permissions([Permission.WRITE_EXTERNAL_STORAGE, Permission.READ_EXTERNAL_STORAGE])
```
How can I obtain read and write permissions for Android in Toga？

**Reply by mhsmith** [reply, marked as answer]
Direct file access with those permissions doesn't work anymore in newer versions of Android, since they introduced ["scoped storage"](https://developer.android.com/training/data-storage#scoped-storage).

Sorry about the `readAllBytes` error: I didn't realize that wasn't introduced until API level 33. I've edited [the example](https://github.com/beeware/toga/pull/1158#issuecomment-1446196973) to use `read` in a loop.

**Reply by wanglifeng95** [reply]
Thanks for your help. The code runs normally, but there is still some confusion

1.  If I want to develop an instant messaging software, I can send messages and pictures from my phone's photo album. If permission requests cannot be issued, I cannot read and send images. How should I use Toga to read album images in newer versions of Android?
2.  The following code can be used normally, I can open the file dialog box and obtain the content of the file. But how do I get the path of the file on my phone?

```
from android.content import Intent
from java import jarray, jbyte

fileChose = Intent(Intent.ACTION_GET_CONTENT)
fileChose.addCategory(Intent.CATEGORY_OPENABLE)
fileChose.setType("*/*")

# Assuming `app` is your toga.App object
results = await app._impl.intent_result(Intent.createChooser(fileChose, "Choose a file"))  
data = results['resultData'].getData()
context = app._impl.native
stream = context.getContentResolver().openInputStream(data)

def read_stream(stream):
    block = jarray(jbyte)(1024 * 1024)
    blocks = []
    while True:
        bytes_read = stream.read(block)
        if bytes_read == -1:
            return b"".join(blocks)
        else:
            blocks.append(bytes(block)[:bytes_read])

content = read_stream(stream)
```
how do I get the path of the selected file?

**Reply by mhsmith** [reply]
> If permission requests cannot be issued, I cannot read and send images. 

Do you need to make a permission request in order to use `ACTION_GET_CONTENT`? You said yourself that "I can open the file dialog box and obtain the content of the file".

> But how do I get the path of the file on my phone?

You can probably extract it from the URL (the `data` variable). But you still won't be able to open it directly.

**Reply by wanglifeng95** [reply]
By converting the URI to a file path, we can use python to read or write the bytes of the files in these folders( 
 '/storage/emulated/0/Android/data or media/packageName/files'  ) or obtain the table data through pandas.read_excel(filePath) and so on.  Currently, I am only testing the situation in these two directories( 
 '/storage/emulated/0/Android/data or media/packageName/files'  ) on my two Android phones, the following methods can obtain the file path of the open_file_dialog box. The code is as follows：

```
async def openFileDialog():
    from toga.platform import current_platform
    print("current_platform : ", current_platform)
    if current_platform == "android":
        from android.content import Intent,ContentResolver
        from android.provider import MediaStore

        context = app._impl.native

        fileChose = Intent(Intent.ACTION_GET_CONTENT)
        fileChose.addCategory(Intent.CATEGORY_OPENABLE)
        fileChose.setType("*/*")

        # Assuming `app` is your toga.App object
        results = await app._impl.intent_result(Intent.createChooser(fileChose, "Choose a file"))
        data = results['resultData'].getData()
        phoneRoot='/storage/emulated/0/'
        dataList=str(data).split(phoneRoot)
        if len(dataList)>1:
            filePath=phoneRoot+dataList[1]
            return filePath

        def getFilePathByURI(uri):
            if uri==None:return None
            scheme=uri.getScheme()
            filePath=None
            if scheme==None:
                filePath=uri.getPath()
            elif ContentResolver.SCHEME_FILE==scheme:
                filePath=uri.getPath()
            elif ContentResolver.SCHEME_CONTENT==scheme:
                cursor = context.getContentResolver().query(uri, None, None, None, None)
                if cursor!=None:
                    if cursor.moveToFirst():
                        columnName = MediaStore.Images.Media.DATA
                        if uri==MediaStore.Images.Media.EXTERNAL_CONTENT_URI:
                            columnName = MediaStore.Images.Media.DATA
                        elif uri==MediaStore.Audio.Media.EXTERNAL_CONTENT_URI:
                            columnName = MediaStore.Audio.Media.DATA
                        elif uri==MediaStore.Video.Media.EXTERNAL_CONTENT_URI:
                            columnName = MediaStore.Video.Media.DATA
                        imageColumnIndex = cursor.getColumnIndex(columnName)
                        if imageColumnIndex<0:
                            filePath=phoneRoot+cursor.getString(0).split(":",1)[1]
                        else:
                            filePath = cursor.getString(imageColumnIndex)
                    cursor.close()
            return filePath

        filePath=getFilePathByURI(data)
    else:
        filePath=await config.app.main_window.open_file_dialog("Open File")
    return filePath

```

I have not fully verified the above code in other situations or versions, so there may be bugs. Therefore, it is for reference only and I hope it can be helpful to others.

**Comment by vzool** [comment]
This code works just fine with Android 15:
```python
async def openFileDialog(self, widget):
    from toga.platform import current_platform
    print("current_platform : ", current_platform)
    if current_platform == "android":
        from android.content import Intent

        fileChose = Intent(Intent.ACTION_GET_CONTENT)
        fileChose.addCategory(Intent.CATEGORY_OPENABLE)
        fileChose.setType("*/*")

        # Assuming `app` is your toga.App object
        results = await self._impl.intent_result(Intent.createChooser(fileChose, "Choose a file"))
        data = results['resultData'].getData()
        context = self._impl.native
        bytesJarray = bytes((context.getContentResolver().openInputStream(data).readAllBytes()))
        # will do something to get filePath and return filePath
       file_path = os.path.join(self.paths.cache, 'import_file')
       with open(file_path, "wb") as file:
           file.write(bytesJarray)
    else:
        filePath = await self.main_window.open_file_dialog("打开文件")
    return filePath
```
How can I use `save_file_dialog` on Android?
Thanks

**Reply by freakboy3742** [reply]
We don't currently support file dialogs on Android. This is at least in part because there is a need to use native File I/O APIs when dealing with files on Android - you can't just use python `open()` calls.

If you want to use file dialogs on Android, you'll need to dig into those native Android File APIs. Unfortunately, I can't point you at a "ready to use" example of this for saving files.

**Reply by mhsmith** [reply]
I've added some more information about saving files at https://github.com/beeware/toga/pull/1158#issuecomment-2254564657.

**Comment by wanglifeng95** [comment]
您好，我已收到来信，会尽快给予你回复。
